File: src/fem_shell/constitutive/failure.py
"""
Failure Criteria for Composite Materials.

This module implements common failure criteria for fiber-reinforced
composite laminae, including:
- Tsai-Wu criterion (quadratic interaction)
- Hashin criterion (mode-specific failure)
- Maximum stress criterion

References
----------
- Tsai, S.W. and Wu, E.M. (1971). "A General Theory of Strength for
  Anisotropic Materials." J. Composite Materials, 5, 58-80.
- Hashin, Z. (1980). "Failure Criteria for Unidirectional Fiber Composites."
  J. Applied Mechanics, 47, 329-334.
"""

from dataclasses import dataclass
from enum import Enum
from typing import Dict, Optional, Tuple

import numpy as np

from fem_shell.core.laminate import StrengthProperties


class FailureMode(Enum):
    """Enumeration of composite failure modes."""

    NO_FAILURE = "no_failure"
    FIBER_TENSION = "fiber_tension"
    FIBER_COMPRESSION = "fiber_compression"
    MATRIX_TENSION = "matrix_tension"
    MATRIX_COMPRESSION = "matrix_compression"
    SHEAR = "shear"
    COMBINED = "combined"  # For Tsai-Wu


@dataclass
class FailureResult:
    """
    Result of failure criterion evaluation.

    Attributes
    ----------
    failed : bool
        True if failure index >= 1
    failure_index : float
        Value of failure criterion (>= 1 indicates failure)
    mode : FailureMode
        Type of failure (for mode-specific criteria)
    margin_of_safety : float
        Safety margin = 1/sqrt(FI) - 1 (for quadratic criteria)
    reserve_factor : float
        Load factor to failure = 1/sqrt(FI)
    """

    failed: bool
    failure_index: float
    mode: FailureMode
    margin_of_safety: float = 0.0
    reserve_factor: float = 1.0

    def __post_init__(self):
        if self.failure_index > 0:
            self.reserve_factor = 1 / np.sqrt(self.failure_index)
            self.margin_of_safety = self.reserve_factor - 1
        else:
            self.reserve_factor = float("inf")
            self.margin_of_safety = float("inf")


def tsai_wu_failure_index(
    sigma: np.ndarray,
    strength: StrengthProperties,
    f12_coefficient: float = -0.5,
) -> FailureResult:
    """
    Compute Tsai-Wu failure index for a composite lamina.

    The Tsai-Wu criterion is a quadratic polynomial criterion:
    F1·σ1 + F2·σ2 + F11·σ1² + F22·σ2² + F66·τ12² + 2F12·σ1·σ2 = 1

    Parameters
    ----------
    sigma : np.ndarray
        Stress in ply principal coordinates [σ1, σ2, τ12] in Pa
    strength : StrengthProperties
        Material strength properties
    f12_coefficient : float, optional
        Coefficient for F12 term. Default is -0.5 (conservative).
        F12 = f12_coefficient / sqrt(Xt·Xc·Yt·Yc)

    Returns
    -------
    FailureResult
        Failure analysis result with index and mode

    Notes
    -----
    The strength parameters are:
    - F1 = 1/Xt - 1/Xc (linear fiber term)
    - F2 = 1/Yt - 1/Yc (linear matrix term)
    - F11 = 1/(Xt·Xc) (quadratic fiber term)
    - F22 = 1/(Yt·Yc) (quadratic matrix term)
    - F66 = 1/S² (shear term)
    - F12 = interaction term (typically -0.5/sqrt(Xt·Xc·Yt·Yc))

    The failure index FI is the value of the left-hand side.
    Failure occurs when FI >= 1.

    Examples
    --------
    >>> strength = StrengthProperties(Xt=1500e6, Xc=1500e6, Yt=40e6, Yc=246e6, S12=68e6)
    >>> sigma = np.array([500e6, 20e6, 30e6])  # [σ1, σ2, τ12]
    >>> result = tsai_wu_failure_index(sigma, strength)
    >>> print(f"FI = {result.failure_index:.3f}, Failed = {result.failed}")
    """
    s1, s2, t12 = sigma

    Xt, Xc = strength.Xt, strength.Xc
    Yt, Yc = strength.Yt, strength.Yc
    S = strength.S12

    # Strength parameters
    F1 = 1 / Xt - 1 / Xc
    F2 = 1 / Yt - 1 / Yc
    F11 = 1 / (Xt * Xc)
    F22 = 1 / (Yt * Yc)
    F66 = 1 / S**2
    F12 = f12_coefficient / np.sqrt(Xt * Xc * Yt * Yc)

    # Failure index
    FI = F1 * s1 + F2 * s2 + F11 * s1**2 + F22 * s2**2 + F66 * t12**2 + 2 * F12 * s1 * s2

    failed = FI >= 1.0

    return FailureResult(
        failed=failed,
        failure_index=FI,
        mode=FailureMode.COMBINED if failed else FailureMode.NO_FAILURE,
    )


def hashin_failure_indices(
    sigma: np.ndarray,
    strength: StrengthProperties,
) -> Dict[FailureMode, FailureResult]:
    """
    Compute Hashin failure indices for all failure modes.

    The Hashin criterion distinguishes between four failure modes:
    1. Fiber tension (σ1 > 0)
    2. Fiber compression (σ1 < 0)
    3. Matrix tension (σ2 > 0)
    4. Matrix compression (σ2 < 0)

    Parameters
    ----------
    sigma : np.ndarray
        Stress in ply principal coordinates [σ1, σ2, τ12] in Pa
    strength : StrengthProperties
        Material strength properties

    Returns
    -------
    Dict[FailureMode, FailureResult]
        Dictionary mapping failure modes to their results

    Notes
    -----
    **Fiber Tension (σ1 ≥ 0):**
    (σ1/Xt)² + (τ12/S)² = 1

    **Fiber Compression (σ1 < 0):**
    |σ1/Xc| = 1

    **Matrix Tension (σ2 ≥ 0):**
    (σ2/Yt)² + (τ12/S)² = 1

    **Matrix Compression (σ2 < 0):**
    (σ2/2ST)² + [(Yc/2ST)² - 1](σ2/Yc) + (τ12/S)² = 1

    where ST is the transverse shear strength.

    Examples
    --------
    >>> strength = StrengthProperties(Xt=1500e6, Xc=1500e6, Yt=40e6, Yc=246e6, S12=68e6)
    >>> sigma = np.array([500e6, -100e6, 30e6])
    >>> results = hashin_failure_indices(sigma, strength)
    >>> for mode, result in results.items():
    ...     print(f"{mode.value}: FI = {result.failure_index:.3f}")
    """
    s1, s2, t12 = sigma

    Xt, Xc = strength.Xt, strength.Xc
    Yt, Yc = strength.Yt, strength.Yc
    S = strength.S12
    ST = strength.S23  # Transverse shear strength

    results = {}

    # Fiber tension (σ1 >= 0)
    if s1 >= 0:
        FI_ft = (s1 / Xt) ** 2 + (t12 / S) ** 2
        results[FailureMode.FIBER_TENSION] = FailureResult(
            failed=FI_ft >= 1.0,
            failure_index=FI_ft,
            mode=FailureMode.FIBER_TENSION,
        )
    else:
        # Fiber compression (σ1 < 0)
        FI_fc = (abs(s1) / Xc) ** 2
        results[FailureMode.FIBER_COMPRESSION] = FailureResult(
            failed=FI_fc >= 1.0,
            failure_index=FI_fc,
            mode=FailureMode.FIBER_COMPRESSION,
        )

    # Matrix tension (σ2 >= 0)
    if s2 >= 0:
        FI_mt = (s2 / Yt) ** 2 + (t12 / S) ** 2
        results[FailureMode.MATRIX_TENSION] = FailureResult(
            failed=FI_mt >= 1.0,
            failure_index=FI_mt,
            mode=FailureMode.MATRIX_TENSION,
        )
    else:
        # Matrix compression (σ2 < 0)
        FI_mc = (s2 / (2 * ST)) ** 2 + ((Yc / (2 * ST)) ** 2 - 1) * (s2 / Yc) + (t12 / S) ** 2
        results[FailureMode.MATRIX_COMPRESSION] = FailureResult(
            failed=FI_mc >= 1.0,
            failure_index=FI_mc,
            mode=FailureMode.MATRIX_COMPRESSION,
        )

    return results


def max_stress_failure_index(
    sigma: np.ndarray,
    strength: StrengthProperties,
) -> FailureResult:
    """
    Compute maximum stress failure criterion.

    The maximum stress criterion compares each stress component
    to its corresponding strength independently (no interaction).

    Parameters
    ----------
    sigma : np.ndarray
        Stress in ply principal coordinates [σ1, σ2, τ12] in Pa
    strength : StrengthProperties
        Material strength properties

    Returns
    -------
    FailureResult
        Failure result with the maximum ratio as failure index

    Notes
    -----
    Failure ratios:
    - R1t = σ1/Xt (fiber tension)
    - R1c = |σ1|/Xc (fiber compression)
    - R2t = σ2/Yt (matrix tension)
    - R2c = |σ2|/Yc (matrix compression)
    - R12 = |τ12|/S (shear)

    FI = max(R1, R2, R12)
    """
    s1, s2, t12 = sigma

    Xt, Xc = strength.Xt, strength.Xc
    Yt, Yc = strength.Yt, strength.Yc
    S = strength.S12

    # Compute individual ratios
    if s1 >= 0:
        R1 = s1 / Xt
        mode1 = FailureMode.FIBER_TENSION
    else:
        R1 = abs(s1) / Xc
        mode1 = FailureMode.FIBER_COMPRESSION

    if s2 >= 0:
        R2 = s2 / Yt
        mode2 = FailureMode.MATRIX_TENSION
    else:
        R2 = abs(s2) / Yc
        mode2 = FailureMode.MATRIX_COMPRESSION

    R12 = abs(t12) / S

    # Find maximum
    ratios = [R1, R2, R12]
    modes = [mode1, mode2, FailureMode.SHEAR]

    max_idx = np.argmax(ratios)
    FI = ratios[max_idx]
    mode = modes[max_idx]

    return FailureResult(
        failed=FI >= 1.0,
        failure_index=FI,
        mode=mode,  # Always return the critical mode
    )


def evaluate_ply_failure(
    sigma: np.ndarray,
    strength: StrengthProperties,
    criterion: str = "tsai-wu",
) -> FailureResult:
    """
    Evaluate ply failure using specified criterion.

    Parameters
    ----------
    sigma : np.ndarray
        Stress in ply principal coordinates [σ1, σ2, τ12] in Pa
    strength : StrengthProperties
        Material strength properties
    criterion : str
        Failure criterion: "tsai-wu", "hashin", or "max-stress"

    Returns
    -------
    FailureResult
        Failure result for the specified criterion.
        For Hashin, returns the most critical mode.
    """
    criterion = criterion.lower().replace("_", "-")

    if criterion == "tsai-wu":
        return tsai_wu_failure_index(sigma, strength)

    elif criterion == "hashin":
        results = hashin_failure_indices(sigma, strength)
        # Return most critical mode
        max_result = max(results.values(), key=lambda r: r.failure_index)
        return max_result

    elif criterion == "max-stress":
        return max_stress_failure_index(sigma, strength)

    else:
        raise ValueError(
            f"Unknown criterion: {criterion}. Use 'tsai-wu', 'hashin', or 'max-stress'."
        )


def stress_transformation_matrix(theta_deg: float) -> np.ndarray:
    """
    Compute stress transformation matrix from laminate to ply coordinates.

    Parameters
    ----------
    theta_deg : float
        Ply angle in degrees

    Returns
    -------
    np.ndarray
        3x3 transformation matrix T such that σ_12 = T @ σ_xy

    Notes
    -----
    Transforms stress from laminate (x,y) coordinates to ply
    principal (1,2) coordinates:

    T = [c²    s²     2sc  ]
        [s²    c²    -2sc  ]
        [-sc   sc   c²-s² ]

    where c = cos(θ), s = sin(θ)
    """
    theta = np.radians(theta_deg)
    c = np.cos(theta)
    s = np.sin(theta)
    c2, s2 = c**2, s**2

    return np.array([[c2, s2, 2 * s * c], [s2, c2, -2 * s * c], [-s * c, s * c, c2 - s2]])


def strain_transformation_matrix(theta_deg: float) -> np.ndarray:
    """
    Compute strain transformation matrix from laminate to ply coordinates.

    Parameters
    ----------
    theta_deg : float
        Ply angle in degrees

    Returns
    -------
    np.ndarray
        3x3 transformation matrix T_eps such that ε_12 = T_eps @ ε_xy

    Notes
    -----
    For engineering strains [εxx, εyy, γxy]:

    T_eps = [c²    s²     sc   ]
            [s²    c²    -sc   ]
            [-2sc  2sc   c²-s²]
    """
    theta = np.radians(theta_deg)
    c = np.cos(theta)
    s = np.sin(theta)
    c2, s2 = c**2, s**2

    return np.array([[c2, s2, s * c], [s2, c2, -s * c], [-2 * s * c, 2 * s * c, c2 - s2]])

File: crates/aeroelast-core/src/materials/failure.rs
//! Failure criteria for fiber-reinforced composite laminae.
//!
//! Implements:
//! - Tsai-Wu criterion (quadratic interaction)
//! - Hashin criterion (mode-specific)
//! - Maximum stress criterion
//!
//! # References
//! - Tsai, S.W. and Wu, E.M. (1971). J. Composite Materials, 5, 58–80.
//! - Hashin, Z. (1980). J. Applied Mechanics, 47, 329–334.

/// Strength properties for a composite lamina.
///
/// All values are in Pa (positive magnitudes — the sign convention is
/// handled internally by each criterion).
#[derive(Debug, Clone)]
pub struct StrengthProperties {
    /// Tensile strength in fiber direction [Pa]
    pub xt: f64,
    /// Compressive strength in fiber direction [Pa]
    pub xc: f64,
    /// Tensile strength transverse to fibers [Pa]
    pub yt: f64,
    /// Compressive strength transverse to fibers [Pa]
    pub yc: f64,
    /// In-plane shear strength [Pa]
    pub s12: f64,
    /// Transverse shear strength [Pa] (defaults to s12/2 if not provided)
    pub s23: f64,
}

impl StrengthProperties {
    pub fn new(xt: f64, xc: f64, yt: f64, yc: f64, s12: f64, s23: Option<f64>) -> Self {
        Self { xt, xc, yt, yc, s12, s23: s23.unwrap_or(s12 / 2.0) }
    }
}

/// Composite failure mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FailureMode {
    NoFailure,
    FiberTension,
    FiberCompression,
    MatrixTension,
    MatrixCompression,
    Shear,
    Combined,
}

/// Result of a failure criterion evaluation.
#[derive(Debug, Clone)]
pub struct FailureResult {
    /// True if failure index >= 1.0
    pub failed: bool,
    /// Failure index value (>= 1.0 means failure)
    pub failure_index: f64,
    /// Failure mode
    pub mode: FailureMode,
    /// Reserve factor = 1 / sqrt(FI)
    pub reserve_factor: f64,
    /// Margin of safety = reserve_factor - 1
    pub margin_of_safety: f64,
}

impl FailureResult {
    pub fn new(failure_index: f64, mode: FailureMode) -> Self {
        let (reserve_factor, margin_of_safety) = if failure_index > 0.0 {
            let rf = 1.0 / failure_index.sqrt();
            (rf, rf - 1.0)
        } else {
            (f64::INFINITY, f64::INFINITY)
        };
        Self {
            failed: failure_index >= 1.0,
            failure_index,
            mode,
            reserve_factor,
            margin_of_safety,
        }
    }
}

/// Compute the Tsai-Wu failure index for a composite lamina.
///
/// Criterion: F1·σ1 + F2·σ2 + F11·σ1² + F22·σ2² + F66·τ12² + 2F12·σ1·σ2 = 1
///
/// # Arguments
/// * `sigma` - `[σ1, σ2, τ12]` stress in ply principal coordinates [Pa]
/// * `strength` - material strength properties
/// * `f12_coefficient` - interaction coefficient (default: -0.5)
pub fn tsai_wu(
    sigma: [f64; 3],
    strength: &StrengthProperties,
    f12_coefficient: f64,
) -> FailureResult {
    let [s1, s2, t12] = sigma;
    let (xt, xc, yt, yc, s) = (strength.xt, strength.xc, strength.yt, strength.yc, strength.s12);

    let f1 = 1.0 / xt - 1.0 / xc;
    let f2 = 1.0 / yt - 1.0 / yc;
    let f11 = 1.0 / (xt * xc);
    let f22 = 1.0 / (yt * yc);
    let f66 = 1.0 / (s * s);
    let f12 = f12_coefficient / (xt * xc * yt * yc).sqrt();

    let fi = f1 * s1 + f2 * s2 + f11 * s1 * s1 + f22 * s2 * s2 + f66 * t12 * t12
        + 2.0 * f12 * s1 * s2;

    let mode = if fi >= 1.0 { FailureMode::Combined } else { FailureMode::NoFailure };
    FailureResult::new(fi, mode)
}

/// Compute Hashin failure indices for all failure modes.
///
/// Returns one `FailureResult` per active mode:
/// - Fiber tension OR fiber compression (depending on sign of σ1)
/// - Matrix tension OR matrix compression (depending on sign of σ2)
///
/// # Arguments
/// * `sigma` - `[σ1, σ2, τ12]` stress in ply principal coordinates [Pa]
/// * `strength` - material strength properties
pub fn hashin(sigma: [f64; 3], strength: &StrengthProperties) -> HashinResults {
    let [s1, s2, t12] = sigma;
    let (xt, xc, yt, yc, s, st) =
        (strength.xt, strength.xc, strength.yt, strength.yc, strength.s12, strength.s23);

    let fiber = if s1 >= 0.0 {
        let fi = (s1 / xt).powi(2) + (t12 / s).powi(2);
        FailureResult::new(fi, FailureMode::FiberTension)
    } else {
        let fi = (s1.abs() / xc).powi(2);
        FailureResult::new(fi, FailureMode::FiberCompression)
    };

    let matrix = if s2 >= 0.0 {
        let fi = (s2 / yt).powi(2) + (t12 / s).powi(2);
        FailureResult::new(fi, FailureMode::MatrixTension)
    } else {
        let fi = (s2 / (2.0 * st)).powi(2)
            + ((yc / (2.0 * st)).powi(2) - 1.0) * (s2 / yc)
            + (t12 / s).powi(2);
        FailureResult::new(fi, FailureMode::MatrixCompression)
    };

    HashinResults { fiber, matrix }
}

/// All Hashin mode results.
#[derive(Debug, Clone)]
pub struct HashinResults {
    pub fiber: FailureResult,
    pub matrix: FailureResult,
}

impl HashinResults {
    /// Returns the most critical result (highest failure index).
    pub fn critical(&self) -> &FailureResult {
        if self.fiber.failure_index >= self.matrix.failure_index {
            &self.fiber
        } else {
            &self.matrix
        }
    }
}

/// Compute the maximum stress failure criterion.
///
/// FI = max(σ1/Xt, |σ1|/Xc, σ2/Yt, |σ2|/Yc, |τ12|/S) depending on sign.
///
/// # Arguments
/// * `sigma` - `[σ1, σ2, τ12]` stress in ply principal coordinates [Pa]
/// * `strength` - material strength properties
pub fn max_stress(sigma: [f64; 3], strength: &StrengthProperties) -> FailureResult {
    let [s1, s2, t12] = sigma;
    let (xt, xc, yt, yc, s) = (strength.xt, strength.xc, strength.yt, strength.yc, strength.s12);

    let (r1, mode1) = if s1 >= 0.0 {
        (s1 / xt, FailureMode::FiberTension)
    } else {
        (s1.abs() / xc, FailureMode::FiberCompression)
    };

    let (r2, mode2) = if s2 >= 0.0 {
        (s2 / yt, FailureMode::MatrixTension)
    } else {
        (s2.abs() / yc, FailureMode::MatrixCompression)
    };

    let r12 = t12.abs() / s;

    let (fi, mode) = [(r1, mode1), (r2, mode2), (r12, FailureMode::Shear)]
        .into_iter()
        .max_by(|a, b| a.0.partial_cmp(&b.0).unwrap())
        .unwrap();

    FailureResult::new(fi, mode)
}

/// Compute the stress transformation matrix from laminate (x,y) → ply (1,2) coordinates.
///
/// T such that σ_12 = T · σ_xy, where σ = [σ_xx, σ_yy, τ_xy].
pub fn stress_transformation(theta_deg: f64) -> [[f64; 3]; 3] {
    let theta = theta_deg.to_radians();
    let (c, s) = (theta.cos(), theta.sin());
    let (c2, s2, sc) = (c * c, s * s, s * c);

    [[c2, s2, 2.0 * sc], [s2, c2, -2.0 * sc], [-sc, sc, c2 - s2]]
}

/// Compute the strain transformation matrix from laminate (x,y) → ply (1,2) coordinates.
///
/// T_eps such that ε_12 = T_eps · ε_xy, where ε = [ε_xx, ε_yy, γ_xy].
pub fn strain_transformation(theta_deg: f64) -> [[f64; 3]; 3] {
    let theta = theta_deg.to_radians();
    let (c, s) = (theta.cos(), theta.sin());
    let (c2, s2, sc) = (c * c, s * s, s * c);

    [[c2, s2, sc], [s2, c2, -sc], [-2.0 * sc, 2.0 * sc, c2 - s2]]
}

#[cfg(test)]
mod tests {
    use super::*;

    fn glass_epoxy() -> StrengthProperties {
        // Typical E-glass/epoxy [Pa]
        StrengthProperties::new(1500e6, 1500e6, 40e6, 246e6, 68e6, None)
    }

    #[test]
    fn tsai_wu_no_failure() {
        let s = glass_epoxy();
        let r = tsai_wu([500e6, 20e6, 30e6], &s, -0.5);
        assert!(r.failure_index < 1.0);
        assert!(!r.failed);
        assert!(r.reserve_factor > 1.0);
    }

    #[test]
    fn tsai_wu_failure() {
        let s = glass_epoxy();
        let r = tsai_wu([1600e6, 0.0, 0.0], &s, -0.5);
        assert!(r.failed);
        assert!(r.failure_index >= 1.0);
    }

    #[test]
    fn hashin_fiber_tension() {
        let s = glass_epoxy();
        let h = hashin([800e6, 5e6, 10e6], &s);
        assert_eq!(h.fiber.mode, FailureMode::FiberTension);
    }

    #[test]
    fn hashin_matrix_compression() {
        let s = glass_epoxy();
        let h = hashin([0.0, -200e6, 30e6], &s);
        assert_eq!(h.matrix.mode, FailureMode::MatrixCompression);
    }

    #[test]
    fn max_stress_shear_critical() {
        let s = glass_epoxy();
        let r = max_stress([10e6, 5e6, 70e6], &s);
        assert_eq!(r.mode, FailureMode::Shear);
        assert!(r.failed);
    }

    #[test]
    fn stress_transformation_identity_at_zero() {
        let t = stress_transformation(0.0);
        // At 0°, T should be identity
        assert!((t[0][0] - 1.0).abs() < 1e-12);
        assert!((t[1][1] - 1.0).abs() < 1e-12);
        assert!((t[2][2] - 1.0).abs() < 1e-12);
        assert!(t[0][1].abs() < 1e-12);
    }

    #[test]
    fn reserve_factor_consistency() {
        let s = glass_epoxy();
        let r = tsai_wu([500e6, 20e6, 30e6], &s, -0.5);
        let expected_rf = 1.0 / r.failure_index.sqrt();
        assert!((r.reserve_factor - expected_rf).abs() < 1e-12);
        assert!((r.margin_of_safety - (expected_rf - 1.0)).abs() < 1e-12);
    }
}
